### PR TITLE
API-2922 Adding the lookup by SubscriptionFieldsId

### PR DESCRIPTION
**Pull request by Tobias1087**

### CODE REVIEW

File: test/it/uk/gov/hmrc/apinotificationqueue/connector/ApiSubscriptionFieldsConnectorSpec.scala [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.apinotificationqueue.connector
+
+;
```
Comment (gokyo) [inline test/it/uk/gov/hmrc/apinotificationqueue/connector/ApiSubscriptionFieldsConnectorSpec.scala]: not needed

File: app/uk/gov/hmrc/apinotificationqueue/controller/QueueController.scala [hunk 2]
```diff
@@ -25,55 +25,67 @@ import play.api.http.HttpEntity
...
 
 import scala.concurrent.Future
 
 @Singleton()
-class QueueController @Inject()(queueService: QueueService, idGenerator: NotificationIdGenerator) extends BaseController {
+class QueueController @Inject()(queueService: QueueService, fieldsService: ApiSubscriptionFieldsService, idGenerator: NotificationIdGenerator) extends BaseController {
 
-  private val CLIENT_ID_HEADER_NAME = "X-Client-ID"
-  private val CLIENT_ID_REQUIRED_ERROR = new BadRequestException(s"$CLIENT_ID_HEADER_NAME required.")
+  val SUBSCRIPTION_FIELD_HEADER_NAME = "api-subscription-fields-id"
```
Comment (gokyo) [inline app/uk/gov/hmrc/apinotificationqueue/controller/QueueController.scala]: why not making these 2 val as `private`?
I understood you prefer to use local constants

File: test/it/uk/gov/hmrc/apinotificationqueue/connector/ApiSubscriptionFieldsConnectorSpec.scala [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+      result.clientId shouldBe clientId
+    }
+
+    "throw an http-verbs Upstream5xxResponse exception if the API Subscription Fields responds with an error" in new Setup {
+      stubFor(get(urlEqualTo(s"/fields/$fieldsId")).willReturn(aResponse().withStatus(500)))
+      intercept[Upstream5xxResponse](await(connector.lookupClientId(fieldsId)))
+    }
+
+    "throw a DocumentationNotFound exception if service is not available in API Subscription Fields" in new Setup {
```
Comment (gokyo) [inline test/it/uk/gov/hmrc/apinotificationqueue/connector/ApiSubscriptionFieldsConnectorSpec.scala]: The correct title should be:
```
throw a `NotFoundException` if the `fieldsId` is not available in the API Subscription Fields service`
```
(or similar)

File: test/unit/uk/gov/hmrc/apinotificationqueue/controller/QueueControllerSpec.scala [hunk 4]
```diff
@@ -51,43 +51,70 @@ class QueueControllerSpec extends UnitSpec with MockitoSugar with WithFakeApplic
...
+
+      status(result) shouldBe Status.BAD_REQUEST
     }
 
-    "throw exception if no payload" in new Setup {
-      val request = FakeRequest(POST, "/queue", Headers(CLIENT_ID_HEADER_NAME -> clientId), AnyContentAsEmpty)
-      intercept[BadRequestException] {
-        await(queueController.save()(request))
-      }
+    "return 400 if api-subscription-fields not there in fieldsService" in new Setup {
```
Comment (gokyo) [inline test/unit/uk/gov/hmrc/apinotificationqueue/controller/QueueControllerSpec.scala]: The title of this test is not very clear to me.
Perhaps you could say 
```
return 400 when the `fieldsId` does not exist in the `api-subscription-fields` service
```
(if this was the scenario you wanted to cover with your test)

File: test/unit/uk/gov/hmrc/apinotificationqueue/controller/QueueControllerSpec.scala [hunk 5]
```diff
@@ -51,43 +51,70 @@ class QueueControllerSpec extends UnitSpec with MockitoSugar with WithFakeApplic
...
+    val mockFieldsService = mock[ApiSubscriptionFieldsService]
+    val queueController = new QueueController(mockQueueService, mockFieldsService, new StaticIDGenerator)
   }
 
   "POST /queue" should {
-    "throw exception when then X-Client-ID header is not sent in the request" in new Setup {
-      intercept[BadRequestException] {
-        await(queueController.save()(FakeRequest(POST, "/queue")))
-      }
+    "return 400 when then X-Client-ID header is not sent in the request" in new Setup {
```
Comment (gokyo) [inline test/unit/uk/gov/hmrc/apinotificationqueue/controller/QueueControllerSpec.scala]: I believe we need to update the title of this test now.
I'd say something like this:
"return 400 when any of the `X-Client-ID` and `api-subscription-fields-id` headers are sent in the request"
Do you agree?

File: app/uk/gov/hmrc/apinotificationqueue/controller/QueueController.scala [hunk 6]
```diff
@@ -25,55 +25,67 @@ import play.api.http.HttpEntity
...
+              )
+            ).map(notification =>
+              Result(ResponseHeader(CREATED, Map(LOCATION -> routes.QueueController.get(notification.notificationId).url)), HttpEntity.NoEntity))
+          }
+
+      })
     }
   }
 
+  private def getClientId(headers: Headers)(implicit hc: HeaderCarrier): Future[Option[String]] = {
```
Comment (gokyo) [inline app/uk/gov/hmrc/apinotificationqueue/controller/QueueController.scala]: This private method is called only inside `save() (when you call `POST /queue`).
Is it correct?
Or shall we use it in all the other endpoints?
If the the `X-Client-ID` header is not provided in the other endpoints, shall we check the `api-subscription-fields-id` header first and then retrieve the client id?

Comment (Tobias1087) [reply app/uk/gov/hmrc/apinotificationqueue/controller/QueueController.scala]: The requirement is only there for the POST everything else is a failure case if not. Possible to move it across them all if needed

Comment (gokyo) [reply app/uk/gov/hmrc/apinotificationqueue/controller/QueueController.scala]: Fair enough.

**Pull request merged**